rest/endpoint/v1/routes.go

```diff
@@ -0,0 +1,175 @@
+	urlQuery := url(prefix, urlPattern, QueryPathFormat, keyspaceParam, tableParam)
+
+	routes := []types.Route{
+		{
+			Method:  http.MethodGet,
+			Pattern: urlColumns,
+			Handler: rl.validateKeyspace(rl.GetColumns),
+		},
+		{
+			Method:  http.MethodPost,
+			Pattern: urlColumns,
+			Handler: rl.validateKeyspace(rl.isSupported(config.TableAlterAdd, rl.AddColumn)),
+		},
+		{
+			Method:  http.MethodDelete,
+			Pattern: urlColumns,
+			Handler: rl.validateKeyspace(rl.isSupported(config.TableAlterDrop, rl.DeleteColumn)),
+		},
+		{
+			Method:  http.MethodGet,
+			Pattern: urlSingleColumn,
+			Handler: rl.validateKeyspace(rl.GetColumn),
+		},
+		{
+			Method:  http.MethodPost,
+			Pattern: urlRows,
+			Handler: rl.validateKeyspace(rl.AddRow),
+		},
+		{
+			Method:  http.MethodGet,
+			Pattern: urlSingleRow,
+			Handler: rl.validateKeyspace(rl.GetRow),
+		},
+		{
+			Method:  http.MethodPut,
+			Pattern: urlSingleRow,
+			Handler: rl.validateKeyspace(rl.UpdateRow),
+		},
+		{
+			Method:  http.MethodDelete,
+			Pattern: urlSingleRow,
+			Handler: rl.validateKeyspace(rl.DeleteRow),
+		},
+		{
+			Method:  http.MethodPost,
+			Pattern: urlQuery,
+			Handler: rl.validateKeyspace(rl.Query),
+		},
+		{
+			Method:  http.MethodGet,
+			Pattern: urlTables,
+			Handler: rl.validateKeyspace(rl.GetTables),
+		},
+		{
+			Method:  http.MethodPost,
+			Pattern: urlTables,
+			Handler: rl.validateKeyspace(rl.isSupported(config.TableCreate, rl.AddTable)),
+		},
+		{
+			Method:  http.MethodGet,
+			Pattern: urlSingleTable,
+			Handler: rl.validateKeyspace(rl.GetTable),
+		},
+		{
+			Method:  http.MethodDelete,
+			Pattern: urlSingleTable,
+			Handler: rl.validateKeyspace(rl.isSupported(config.TableDrop, rl.DeleteTable)),
+		},
+		{
+			Method:  http.MethodGet,
+			Pattern: urlKeyspaces,
+			Handler: http.HandlerFunc(rl.GetKeyspaces),
+		},
+	}
+
+	return routes
+}
+
+func url(prefix string, urlPattern config.UrlPattern, format string, parameterNames ...string) string {
+	return path.Join(prefix, urlPattern.UrlPathFormat(format, parameterNames...))
+}
+
+func (s *routeList) validateKeyspace(next http.HandlerFunc) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		keyspaceName := s.params(r, keyspaceParam)
+
+		if s.singleKeyspace != "" && s.singleKeyspace != keyspaceName {
+			// Only a single keyspace is allowed and it's not the provided one
+			RespondWithKeyspaceNotAllowed(w)
+			return
+		}
+
+		if s.excludedKeyspaces[keyspaceName] {
+			RespondWithKeyspaceNotAllowed(w)
+			return
+		}
+
+		next(w, r)
+	}
+}
+
+func (s *routeList) isSupported(requiredOp config.SchemaOperations, handler http.HandlerFunc) http.HandlerFunc {
+	if s.operations.IsSupported(requiredOp) {
+		return handler
+	}
+
+	return forbiddenHandler
+}
+
+func forbiddenHandler(w http.ResponseWriter, _ *http.Request) {
+	http.Error(w, http.StatusText(http.StatusInternalServerError),
```

Comment:
`http.StatusMethodNotAllowed`  might make more sense here? `http.StatusInternalServerError` makes it seem like there was an unrecoverable error processing the request.

Reply:
good catch (I was copying and pasting).